docs/internal/rfc/009rfc-transition-explorer.md

```diff
@@ -0,0 +1,459 @@
+[TransitionExecutor
+class](https://github.com/informalsystems/apalache/tree/master/tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/trex/TransitionExecutor.scala).
+During previous iterations of the proposed API we discussed exposing a
+higher-level API, targeted at meeting the requirements more directly.  However,
+discussion revealed that the expensive computational costs of SAT solving in
+most cases made it infeasible to meet the requirements in this way.  Instead, we
+must expose most of the underlying logic of the `TransitionExecutor`, and task
+the users with building their own exploration strategies with these primitives.
+
+It is likely that we will be able to provide some higher-level functionality to users
+by way of wrapper libraries we implement on top of the proposed API, but that
+work should be left to a subsequent iteration.
+
+**NOTE:** This interface is intended as an abstract API, to communicate the
+mappings from request to reply. See the [gRPC Example](#grpc-example) for a
+sketch of what the actual implementation may look like.
+
+```scala
+/** A State is a map from variable names to values, and represents a concrete state.  */
+type State = Map[String, TlaEx]
+
+/** An abstract representation of a transition.
+ *
+ * These correspond to the numbered transitions that the `TransitionExectur` uses
+ * to advance frames. But we likely want to present some more illuminating metadata
+ * associated with the transitions to the users. E.g., ability to view a
+ * representation as a `TlaEx` or see an associated operator name (if any)?  */
+type Transition
+
+/** An execution is an alternating sequence of states and transitions,
+ *  terminating in a state with no transition.
+ *
+ * It is a high-level representation of the `EncodedExecution` maintained by the
+ * transition executor).
+ *
+ * E.g., an execution from `s1` to `sn` with transitions `ti` for each `i` in `1..n-1`:
+ * 
+ *     List((s1, Some(t1)), (s2, Some(t2)), ..., (sn, None))
+ */
+type Execution = List[(State, Option[Transition])]
+
+trait UninitializedModel = {
+  val spec: TlaModule
+  val transitions: Set(Transition)
+}
+
+trait InitializedModel extends UninitializedModel {
+  val constInitPrimed: Map[String, TlaEx]
+}
+
+/** An abstract representation of the `CheckerInput`
+ *
+ * A `Model` includes all static data representing a model.
+ * 
+ * An `UninitializedModel` is missing the information that would be needed in
+ * order to actually explore its symbolic states (such as the initial values of
+ * its constants).
+ *
+ * An `InitializedModel` has all data needed to explore its symbolic states. */
+type Model = Either[UninitializedModel, InitializedModel]
+
+/**  The type of errors that can result from failures when loading a spec. */
+type LoadErr
+
+/**  The type of errors that can result from failures to assert a constraint. */
+type AssertErr
+
+/**  The type of errors that can result from checking satisfiability of a frame. */
+type SatError
```

Comment:
Just curious, are you going to use case classes for these errors? It's not clear to me why we want to define error types like that.

Reply:
Yes, I think so. Only it's not totally clear yet to me whether we'll want different error types (allow to create distinct parts of the computation, narrowing the kind of error handling required in those segments) or if they should all just be sub-types of a single error type.